src/lsm/block_free_set.zig

```diff
@@ -1,37 +1,421 @@
+    try expectEqual(@as(?u64, null), set.acquire());
+
+    i = 0;
+    while (i < blocks_count) : (i += 1) set.release(@as(u64, i + 1));
+    try expect_block_free_set_equal(empty, set);
+
+    i = 0;
+    while (i < blocks_count) : (i += 1) try expectEqual(@as(?u64, i + 1), set.acquire());
+    try expectEqual(@as(?u64, null), set.acquire());
+
+    // Exercise the RingBuffer-index sync by de/re-allocating the last bit of a shard.
+    set.release(i);
+    try expectEqual(@as(usize, 1), set.recent.count);
+    try expectEqual(@as(?u64, i), set.acquire());
+    try expectEqual(@as(usize, 0), set.recent.count);
+    try expectEqual(@as(?u64, null), set.acquire());
+}
+
+fn test_block_shards_count(expect_shards_count: usize, blocks_count: usize) !void {
+    var set = try BlockFreeSet.init(std.testing.allocator, blocks_count);
+    defer set.deinit(std.testing.allocator);
+
+    try std.testing.expectEqual(expect_shards_count, set.index.bit_length);
+}
+
+test "BlockFreeSet encode, decode, encode" {
+    const shard_size = BlockFreeSet.shard_size / @bitSizeOf(usize);
+    // Uniform.
+    try test_encode(&.{.{.fill = .uniform_ones, .words = shard_size}});
+    try test_encode(&.{.{.fill = .uniform_zeros, .words = shard_size}});
+    try test_encode(&.{.{.fill = .literal, .words = shard_size}});
+    try test_encode(&.{.{.fill = .uniform_ones, .words = std.math.maxInt(u16) + 1}});
+
+    // Mixed.
+    try test_encode(&.{
+        .{.fill = .uniform_ones, .words = shard_size / 4},
+        .{.fill = .uniform_zeros, .words = shard_size / 4},
+        .{.fill = .literal, .words = shard_size / 4},
+        .{.fill = .uniform_ones, .words = shard_size / 4},
+    });
+
+    // Random.
+    var seed: u64 = undefined;
+    try std.os.getrandom(mem.asBytes(&seed));
+    var prng = std.rand.DefaultPrng.init(seed);
+
+    const fills = [_]TestPatternFill{.uniform_ones, .uniform_zeros, .literal};
+    var t: usize = 0;
+    while (t < 10) : (t += 1) {
+        var patterns = std.ArrayList(TestPattern).init(std.testing.allocator);
+        defer patterns.deinit();
+        var i: usize = 0;
+        while (i < shard_size) : (i += 1) {
+            try patterns.append(.{
+                .fill = fills[prng.random.uintLessThan(usize, fills.len)],
+                .words = 1,
+            });
+        }
+        try test_encode(patterns.items);
+    }
+}
+
+const TestPattern = struct {
+    fill: TestPatternFill,
+    words: usize,
+};
+
+const TestPatternFill = enum { uniform_ones, uniform_zeros, literal };
+
+fn test_encode(patterns: []const TestPattern) !void {
+    var seed: u64 = undefined;
+    try std.os.getrandom(mem.asBytes(&seed));
+    var prng = std.rand.DefaultPrng.init(seed);
+
+    var blocks_count: usize = 0;
+    for (patterns) |pattern| blocks_count += pattern.words * @bitSizeOf(usize);
+    var decoded_expect = try BlockFreeSet.init(std.testing.allocator, blocks_count);
```

Comment:
```suggestion

    var decoded_expect = try BlockFreeSet.init(std.testing.allocator, blocks_count);
```

Just to make the allocation/deallocation clear and grouped.